Data/ByteString/Lazy.hs

```diff
@@ -571,29 +588,25 @@ compareLength (Chunk c cs) toCmp  = compareLength cs (toCmp - fromIntegral (S.le
    n <= length t = compareLength t n /= LT
   #-}
 
+mapAccumLChunks :: (acc -> S.ByteString -> (acc, S.ByteString)) -> acc -> ByteString -> (acc, ByteString)
+mapAccumLChunks function accumulator = fmap (List.foldr Chunk Empty) . List.mapAccumL function accumulator . foldrChunks (:) [ ]
+
+mapAccumRChunks :: (acc -> S.ByteString -> (acc, S.ByteString)) -> acc -> ByteString -> (acc, ByteString)
+mapAccumRChunks function accumulator = fmap (List.foldr Chunk Empty) . List.mapAccumR function accumulator . foldrChunks (:) [ ]
+
 -- | The 'mapAccumL' function behaves like a combination of 'map' and
 -- 'foldl'; it applies a function to each element of a ByteString,
 -- passing an accumulating parameter from left to right, and returning a
 -- final value of this accumulator together with the new ByteString.
 mapAccumL :: (acc -> Word8 -> (acc, Word8)) -> acc -> ByteString -> (acc, ByteString)
-mapAccumL f = go
-  where
-    go s Empty        = (s, Empty)
-    go s (Chunk c cs) = (s'', Chunk c' cs')
-        where (s',  c')  = S.mapAccumL f s c
-              (s'', cs') = go s' cs
+mapAccumL = mapAccumLChunks . S.mapAccumL
 
 -- | The 'mapAccumR' function behaves like a combination of 'map' and
 -- 'foldr'; it applies a function to each element of a ByteString,
 -- passing an accumulating parameter from right to left, and returning a
 -- final value of this accumulator together with the new ByteString.
 mapAccumR :: (acc -> Word8 -> (acc, Word8)) -> acc -> ByteString -> (acc, ByteString)
-mapAccumR f = go
-  where
-    go s Empty        = (s, Empty)
-    go s (Chunk c cs) = (s'', Chunk c' cs')
-        where (s'', c') = S.mapAccumR f s' c
-              (s', cs') = go s cs
+mapAccumR = mapAccumRChunks . S.mapAccumR
```

Comment:
Do these refactorings change anything important about these functions, e.g. strictness properties, performance, inlining behaviour?

(Apologies if I've missed earlier discussions on this.)

Reply:
I do not think there was any discussion.

For some reason there are not any bench marks for these functions. I went on to add some and compare, and it turns out that the previous versions run somewhat faster on short inputs _(up to 16 bytes in length)_. However, they perform the same on larger inputs. It looks like the constant addend of the time complexity formula has grown by about 80 nanoseconds. Switching optimizations off makes it about 11 times slowly uniformly across all sizes, so it is safe to say that asymptotic complexity of the algorithm itself is unchanged.

I cannot say anything about strictness right now but I can put my strictness check draft to work on it if you would like me to. It might take a few days.

I cannot say anything about inlining at all.

Reply:
Thanks for checking! :)

Could you clarify what the motivation for the refactorings was? Asked differently, what would we lose if we'd revert the refactoring?

Reply:
This is a great question.

In what follows, I am going to discuss the _«right»_ functions, but the same applies to the _«left»_ ones.

So, at first we wanted to write both `scanr` and `mapAccumR` in terms of `mapAccumRChunks` in order to share the quality assurance already in place for `mapAccumR` with `scanr`. But it turned out that we can write `scanr` in terms of `mapAccumR` itself and `mapAccumRChunks` is not needed. I did not notice that until now.

On the good side, `mapAccumRChunks` leverages the natural isomorphose _(up to strictness)_ of byte stream and list of bytes. This means that we export responsibility to `base`: if their definitions have some good properties, our definitions also would. We then propagate these good properties down to `mapAccumR` and `scanr`. So, from the view of quality assurance, it is an improvement.

The conclusion is that by reverting this change we would lose some quality assurance. `mapAccumR` was written back in 2007 in 130906ef928f9761978dd7d13bd4b3082badd1b1 and the only thing that changed since then is some η reduction, so we can be reasonably sure that it works fine already. On the other hand, by keeping this change we expose ourselves to unknown dangers. I am going to revert this change and throw away `mapAccumRChunks`.